crates/redis/src/lib.rs

```diff
@@ -52,30 +56,78 @@ impl TriggerExecutor for RedisTrigger {
     type RunConfig = NoArgs;
 
     async fn new(engine: TriggerAppEngine<Self>) -> Result<Self> {
-        let address = engine
+        let default_address: String = engine
             .trigger_metadata::<TriggerMetadata>()?
             .unwrap_or_default()
             .address;
 
-        let mut channel_components: HashMap<String, Vec<String>> = HashMap::new();
+        let mut addresses = HashSet::new();
+
+        let mut server_channels: HashMap<String, ChannelComponents> = HashMap::new();
 
         for (_, config) in engine.trigger_configs() {
-            channel_components
+            let address = config.address.clone().unwrap_or(default_address.clone());
+            addresses.insert(address.clone());
+            let server = server_channels.entry(address).or_default();
+            server
                 .entry(config.channel.clone())
                 .or_default()
                 .push(config.component.clone());
         }
         Ok(Self {
-            engine,
-            address,
-            channel_components,
+            engine: Arc::new(engine),
+            addresses: addresses.into_iter().collect(),
+            server_channels,
         })
     }
 
     /// Run the Redis trigger indefinitely.
     async fn run(self, _config: Self::RunConfig) -> Result<()> {
-        let address = &self.address;
+        let addresses = self.addresses.clone();
+        let tasks: Vec<_> = addresses
+            .into_iter()
+            .map(|address| {
+                let trigger = self.clone();
+                tokio::spawn(async move { trigger.run_listener(address).await })
+            })
+            .collect();
+
+        // wait for the first handle to be returned and drop the rest
+        let (_, _, rest) = futures::future::select_all(tasks).await;
+        drop(rest);
+        println!("exiting");
```

Comment:
Is this needed?  Do other triggers print it?

Reply:
Other triggers do not. I will remove it.